Add common library

alvarocabanas opened the pull request.

roobre commented:
i think having the chart moved from the old repo and updated to use the library in the same PR makes review a bit harder as there's no diff to check, everything is an addition.
Do you think it would make sense to move the chart as-is first (so that PR would require no review), and then make this PR on top?

alvarocabanas commented:
> i think having the chart moved from the old repo and updated to use the library in the same PR makes review a bit harder as there's no diff to check, everything is an addition. Do you think it would make sense to move the chart as-is first (so that PR would require no review), and then make this PR on top?



> i think having the chart moved from the old repo and updated to use the library in the same PR makes review a bit harder as there's no diff to check, everything is an addition. Do you think it would make sense to move the chart as-is first (so that PR would require no review), and then make this PR on top?

I created this branch From [helm-chart-repo](https://github.com/newrelic/newrelic-k8s-metrics-adapter/tree/feature/helm-chart-repo) and that is the one adding the chart, I will create the PR to that branch so you can propperly see the changes.

alvarocabanas commented:
> I have merged the Helm chart into `main`. As you used my branch as a base maybe you would like to rebase from `main` so it is easier to review...
> 
> I am not able to review it: <img alt="Screenshot 2022-03-15 at 16 29 47" width="160" src="https://user-images.githubusercontent.com/53659978/158413156-f33e3db5-84bf-4aaa-ab88-daff98af6b16.png">

Sure, tomorrow morning I'll do it in this and the helm-docs ;)

paologallinaharbur commented on .github/workflows/ci.yaml:
```diff
@@ -107,13 +107,17 @@ jobs:
           NEWRELIC_ACCOUNT_ID: "${{ secrets.NEWRELIC_ACCOUNT_ID }}"
         run: |
           make -C newrelic-k8s-metrics-adapter generate-local-values
+          helm dependency build ./helm-charts-newrelic/charts/newrelic-k8s-metrics-adapter
```
Why are we updated dependencies for  `helm-charts-newrelic` chart?

We should have it locally and that chart has no dependencies This should not be needed, right?

alvarocabanas replied:
It's needed in order to get the common-library

alvarocabanas replied:
If not it was failing

kang-makes commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/job-patchAPIService.yaml:
```diff
@@ -3,50 +3,58 @@ apiVersion: batch/v1
+        app: {{ template "common.naming.name" $ }}
+{{ include "common.labels" $ | indent 8 }}
     spec:
       {{- if .Values.apiServicePatchJob.image.pullSecrets }}
       imagePullSecrets:
         {{- toYaml .Values.apiServicePatchJob.image.pullSecrets | nindent 8 }}
       {{- end }}
```
Is there any issue that does not allow us to use a global pull secrets here?

alvarocabanas replied:
Yes, I found it weird that we had 2 separated ones, maybe @gsanchezgavier or @paologallinaharbur can clarify why it was thought having 2 different pullsecrets.

alvarocabanas replied:
Fixed in last commits

kang-makes commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/job-patchAPIService.yaml:
```diff
@@ -3,50 +3,58 @@ apiVersion: batch/v1
       restartPolicy: OnFailure
-      serviceAccountName: {{ template "newrelic-k8s-metrics-adapter.fullname" . }}-apiservice
+      serviceAccountName: {{ include "common.serviceAccount.name" . }}-apiservice
       securityContext:
         runAsGroup: 2000
         runAsNonRoot: true
         runAsUser: 2000
```
[This PR is still pending](https://github.com/newrelic/helm-charts/pull/725).

alvarocabanas replied:
Same as the one above

kang-makes commented on charts/newrelic-k8s-metrics-adapter/templates/configmap.yaml:
```diff
@@ -1,9 +1,9 @@
+    {{- include "common.labels" . | nindent 4 }}
 data:
   config.yaml: |
     accountID: {{ .Values.config.accountID | required "config.accountID is required" }}
```
Does make sense to have this as a global too?

alvarocabanas replied:
@gsanchezgavier, @paologallinaharbur ?

paologallinaharbur replied:
As far as I know it is the only chart we own with that. StatsD has it and no idea about pixie.

Considering that we would need to implement it and then change chart we do not own, I would not say it is not really needed

alvarocabanas replied:
oks, we agreed to not add it as a global for the reasons exposed above.

kang-makes commented on charts/newrelic-k8s-metrics-adapter/templates/deployment.yaml:
```diff
@@ -18,22 +18,22 @@ spec:
-      serviceAccountName: {{ template "newrelic-k8s-metrics-adapter.serviceAccountName" . }}
+      serviceAccountName: {{ include "common.serviceAccount.name" . }}
       {{- if .Values.podSecurityContext.enabled }}
       securityContext:
         runAsUser: {{ .Values.podSecurityContext.runAsUser }}
         runAsGroup: {{ .Values.podSecurityContext.runAsGroup }}
         fsGroup: {{ .Values.podSecurityContext.fsGroup }}
```
[This PR is still pending](https://github.com/newrelic/helm-charts/pull/725).

alvarocabanas replied:
Fixed in last version

kang-makes commented on charts/newrelic-k8s-metrics-adapter/templates/deployment.yaml:
```diff
@@ -55,38 +55,50 @@ spec:
-        {{- toYaml .Values.affinity | nindent 8 }}
+        {{- . | nindent 8 }}
       {{- end }}
       {{- if $.Values.hostNetwork }}
```
[This PR is still pending](https://github.com/newrelic/helm-charts/pull/750).

alvarocabanas replied:
Fixed

kang-makes commented on charts/newrelic-k8s-metrics-adapter/tests/configmap_test.yaml:
```diff
@@ -0,0 +1,92 @@
+suite: test configmap region helper and externalMetrics
+templates:
+  - templates/configmap.yaml
+release:
+  name: my-release
+  namespace: my-namespace
+tests:
+  - it: has the correct region when defined in local values
+    set:
+      personalAPIKey: 21321
+      config:
+        accountID: 111
+        region: A-REGION
```
If they are needed you can put them in the header so you don't bother about them and have to repeat them.
```suggestion
release:
  name: my-release
  namespace: my-namespace
set:
  personalAPIKey: 21321
  config:
    accountID: 111
tests:
  - it: has the correct region when defined in local values
    set:
      config:
        region: A-REGION
```

alvarocabanas replied:
It's not working, actually in the tests of the common library they are not doing anything.

kang-makes commented on charts/newrelic-k8s-metrics-adapter/tests/deployment_test.yaml:
```diff
@@ -0,0 +1,134 @@
+          value:
+            disktype: ssd
+        template: templates/deployment.yaml
+  - it: renders affinity
```
Affinity comes [from the common-library](https://github.com/newrelic/helm-charts/blob/master/library/common-library/templates/_affinity.tpl) and [it is tested](https://github.com/newrelic/helm-charts/blob/master/library/CHART-TEMPLATE/tests/common_library_affinity_test.yaml). Is there any reason to re-test it?

alvarocabanas replied:
It's to test they get rendered, if the template has a problem the library won't render them, whenever there is an if, I like to put a test with the happy path.

alvarocabanas replied:
same as the other, only added to one test with all the values ensuring it gets rendered.

kang-makes commented on charts/newrelic-k8s-metrics-adapter/tests/job_patch_job_patchapiservice_test.yaml:
```diff
@@ -0,0 +1,69 @@
+    asserts:
+      - hasDocuments:
+          count: 1
+  - it: tolerations are correctly rendered
```
Tolerations come [from the common-library](https://github.com/newrelic/helm-charts/blob/master/library/CHART-TEMPLATE/tests/common_library_tolerations_test.yaml) and [they are tested there](https://github.com/newrelic/helm-charts/blob/master/library/CHART-TEMPLATE/tests/common_library_tolerations_test.yaml).

alvarocabanas replied:
yes, but here it's just tested that they get rendered.

alvarocabanas replied:
If the template is having an issue they would not get rendered.

alvarocabanas replied:
I've modified it to not be specifically tested, but just added to the vaules

kang-makes commented on charts/newrelic-k8s-metrics-adapter/ci/test-values.yaml:
```diff
@@ -6,6 +6,7 @@ config:
 
 serviceAccount:
   name: test
+  create: true
```
Or is it for the CI to pass?

paologallinaharbur replied:
Was this supposed to be marked as resolved? I was expecting a commit removing that

alvarocabanas replied:
There is a commit with that https://github.com/newrelic/newrelic-k8s-metrics-adapter/pull/41/commits/d31c4526ea9a292b722ea94351b4f8598e53854a

kang-makes commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/job-createSecret.yaml:
```diff
@@ -3,48 +3,56 @@ apiVersion: batch/v1
       restartPolicy: OnFailure
-      serviceAccountName: {{ template "newrelic-k8s-metrics-adapter.fullname" . }}-apiservice
+      serviceAccountName: {{ include "common.serviceAccount.name" . }}-apiservice
       securityContext:
         runAsGroup: 2000
         runAsNonRoot: true
         runAsUser: 2000
```
Pending PR from the common-library.

alvarocabanas replied:
This chart has 2 different podSecurityContexts, the one in the deployment is coming from the values using the helper, but this one we have not been letting the users modify it and can't be in the values with the format expected by the helper since it's already used by the deployment one.

paologallinaharbur replied:
No idea if it would make sense to have an other value for that..  😕 
Maybe we can keep that fixed for now, no strong opinion

Using the "same" security context could lead to issues if at some point we need them to be different

alvarocabanas replied:
I agree, I would leave it non modifiable in this case, WDYT Juanma?

alvarocabanas replied:
oks, I keep it fixed for now

kang-makes reviewed (commented):
Impressive work. Everything missing or is it missing from `common-library` or are nits.

While we finish the common library I left a few comments.

There are a bunch of objects that has no 
```yaml
metadata:
  namespace: {{ .Release.Namespace }}
```

Would be awesome for things like the guided install that templates these charts into a manifest or tests like `helm template | kubectl apply -f -` as this chart uses more than one ns and putting `kubectl -n ns` will yell that there are objects that does not match the namespace.

Great work :)

alvarocabanas commented on charts/newrelic-k8s-metrics-adapter/Chart.yaml:
```diff
@@ -8,6 +8,12 @@ sources:
+dependencies:
+  - name: common-library
+    version: 0.9.0
+    repository: "https://helm-charts.newrelic.com"
```
Everywhere we are using the wrapper, also on the guided install, I wouldn't give 2 different informations, WDYT?

kang-makes replied:
Cool then. I mark it a solved.

alvarocabanas commented on charts/newrelic-k8s-metrics-adapter/templates/_helpers.tpl:
```diff
@@ -77,29 +15,14 @@ Return the licenseKey
-  {{- if .Values.global.nrStaging -}}
+  {{- if (include "common.nrStaging" .) -}}
     Staging
   {{- else if eq (include "newrelic-k8s-metrics-adapter.licenseKey" . | substr 0 2) "eu" -}}
```
Changed to the helper

kang-makes commented on charts/newrelic-k8s-metrics-adapter/templates/deployment.yaml:
```diff
@@ -100,6 +104,6 @@ spec:
-      {{- if $.Values.hostNetwork }}
-      hostNetwork: {{ $.Values.hostNetwork }}
+      {{- with include "common.hostNetwork" . }}
+      hostNetwork: {{ . }}
```
As we talked in the stand up just leaving it here as a reminder because you are switching context) we now have a [new helper that returns the value directly](https://github.com/newrelic/helm-charts/blob/cd27a2262b8b953a7e5f921de89f8c81f3ead7c4/library/common-library/templates/_hostnetwork.tpl#L31-L42)
There is a couple of examples with a [configMap](https://github.com/newrelic/helm-charts/blob/cd27a2262b8b953a7e5f921de89f8c81f3ead7c4/library/CHART-TEMPLATE/templates/example-cm-hostnetwork.yaml#L9-L12) and a [Deployment](https://github.com/newrelic/helm-charts/blob/cd27a2262b8b953a7e5f921de89f8c81f3ead7c4/library/CHART-TEMPLATE/templates/deployment.yaml#L38).

alvarocabanas replied:
I see it's still not merged, so I leave it to change once it is.

alvarocabanas replied:
Solved in last commit

kang-makes reviewed (commented):
A couple of things that still need a bit of love and I would approve it (after another deep review because it is huge).

paologallinaharbur commented on charts/newrelic-k8s-metrics-adapter/templates/_helpers.tpl:
```diff
@@ -24,10 +9,10 @@ Select a value for the region
   {{- if (include "common.nrStaging" .) -}}
     Staging
-  {{- else if eq (include "newrelic-k8s-metrics-adapter.licenseKey" . | substr 0 2) "eu" -}}
+  {{- else if eq (include "common.license._licenseKey" . | substr 0 2) "eu" -}}
     EU
     {{- end }}
-{{- else if eq (include "newrelic-k8s-metrics-adapter.licenseKey" . | substr 0 2) "eu" -}}
+{{- else if eq (include "common.license._licenseKey" . | substr 0 2) "eu" -}}
 EU
 {{- end -}}
```
isn't .Values.config.region?

alvarocabanas replied:
it is

alvarocabanas commented on charts/newrelic-k8s-metrics-adapter/templates/_helpers.tpl:
```diff
@@ -99,12 +7,12 @@ Select a value for the region
     EU
     {{- end }}
-{{- else if eq (include "newrelic-k8s-metrics-adapter.licenseKey" . | substr 0 2) "eu" -}}
+{{- else if eq (include "common.license._licenseKey" . | substr 0 2) "eu" -}}
```
Regarding the not returning US is because if no region is passed, then will always be US.
Regarding the other 2 points, you are right and I corrected it.

alvarocabanas commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/job-createSecret.yaml:
```diff
@@ -3,48 +3,56 @@ apiVersion: batch/v1
   labels:
-    app: {{ template "newrelic-k8s-metrics-adapter.name" $ }}
-{{ include "newrelic-k8s-metrics-adapter.labels" $ | indent 4 }}
+    app: {{ template "common.naming.name" $ }}
```
no, it's redundant I will remove it.

alvarocabanas replied:
done

paologallinaharbur commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/psp.yaml:
```diff
@@ -2,13 +2,13 @@
-    app: {{ template "newrelic-k8s-metrics-adapter.name" . }}
-{{ include "newrelic-k8s-metrics-adapter.labels" . | indent 4 }}
+    app: {{ template "common.naming.name" . }}
+{{ include "common.labels" . | indent 4 }}
```
I saw it in different places, but I would prefer to have it as:

```
  labels:
    app: {{ template "common.naming.name" . }}
    {{- include "common.labels" . | nindent 4 -}}
```

In this was it is indented correctly automatically

(NIT about style, can be ignored)

alvarocabanas replied:
I agree about you, I didn't fix it and it's a good nit.

alvarocabanas replied:
done

paologallinaharbur commented on charts/newrelic-k8s-metrics-adapter/templates/configmap.yaml:
```diff
@@ -1,9 +1,9 @@
+  name: {{ template "common.naming.fullname" . }}
   labels:
-    {{- include "newrelic-k8s-metrics-adapter.labels" . | nindent 4 }}
+    {{- include "common.labels" . | nindent 4 }}
```
like this one!

alvarocabanas replied:
done

paologallinaharbur commented on charts/newrelic-k8s-metrics-adapter/templates/deployment.yaml:
```diff
@@ -18,22 +18,20 @@ spec:
-      {{- if .Values.podSecurityContext.enabled }}
+      serviceAccountName: {{ include "common.serviceAccount.name" . }}
+      {{- with include "common.securityContext.pod" . }}
       securityContext:
```
I see common.securityContext.pod, do we need as well common.securityContext.container?

alvarocabanas replied:
Originally, this chart didn't allow it, do you think we should add it as an extra feature?

kang-makes replied:
Yes. The same applies to things like `hostNetwork` and every aspect possible that is in the common library and makes sense.

alvarocabanas replied:
I added it.

paologallinaharbur reviewed (commented):
massive improvement 😄 

Some decision are still pending but I like the current shape (I did not check tests yet)!

alvarocabanas commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/job-createSecret.yaml:
```diff
@@ -3,48 +3,56 @@ apiVersion: batch/v1
+      name: {{ template "common.naming.fullname" . }}-apiservice-create
       labels:
-        app: {{ template "newrelic-k8s-metrics-adapter.name" $ }}
-{{ include "newrelic-k8s-metrics-adapter.labels" $ | indent 8 }}
```
Yes, I have just added them

paologallinaharbur commented on charts/newrelic-k8s-metrics-adapter/templates/deployment.yaml:
```diff
@@ -55,39 +57,49 @@ spec:
       {{- end }}
-      {{- if .Values.affinity }}
+      {{- with include "common.affinity" . }}
       affinity:
```
should we add as well the [dns](https://github.com/newrelic/helm-charts/blob/master/library/common-library/templates/_dnsconfig.tpl) helper?
should we add [fedramp](https://github.com/newrelic/helm-charts/blob/master/library/common-library/templates/_fedramp.tpl)?
should we add the [proxy](https://github.com/newrelic/helm-charts/blob/master/library/common-library/templates/_proxy.tpl)? @marcsanmi

marcsanmi replied:
I'm not sure yet if adding the proxy will work.

alvarocabanas replied:
I added Dns, fedramp is not used here, and proxy will be added in the proxy feature marc is working on.

paologallinaharbur commented on charts/newrelic-k8s-metrics-adapter/tests/job_patch_common_test.yaml:
```diff
@@ -0,0 +1,26 @@
+suite: test labels and renderingfor job-batch objects
```
rendering[ ]for

alvarocabanas replied:
Fixed, thanks

paologallinaharbur reviewed (approved):
Some helpers can be still added. But all good so far

kang-makes commented on charts/newrelic-k8s-metrics-adapter/values.yaml:
```diff
@@ -62,18 +63,22 @@ resources:
 # Configure podSecurityContext
 podSecurityContext:
-  enabled: false
   fsGroup: 1001
   runAsUser: 1001
   runAsGroup: 1001
```
This overrides the one that comes from `.global`. This is the reason I has a `defaults` [helper](https://github.com/newrelic/nri-kubernetes/pull/383/commits/06f77ab4536590f58d1ea673b48365e678805742#diff-3dc1d0b38b4813f0520e3397a987711789b05548c1fc55680d8bee2f3bc18699R20-R26) and [removed from the Values](https://github.com/newrelic/nri-kubernetes/pull/383/commits/06f77ab4536590f58d1ea673b48365e678805742#diff-b1c831b81fe52267808d0d6996f50be42192cf1afbf1cd8502d6313db9ed8f67L384-L393) in the `nri-kubernetes` helm chart.

alvarocabanas replied:
Fixed

kang-makes reviewed (commented):
We are almost done. Hang in there!

kang-makes commented on .github/workflows/ci.yaml:
```diff
@@ -64,6 +64,9 @@ jobs:
           fetch-depth: 0
       - uses: helm/chart-testing-action@v2.2.1
 
+      - name: Add newrelic repository
+        run: helm repo add newrelic https://helm-charts.newrelic.com
+
```
Do we really need this? Every chart installed during linting/integration is local. This is the only pipeline where this is failing :/

kang-makes replied:
I am able to reproduce it and I don't know why this happens... Maybe another Helm issue.

kang-makes commented on charts/newrelic-k8s-metrics-adapter/Chart.lock:
```diff
@@ -0,0 +1,6 @@
+dependencies:
+- name: common-library
+  repository: https://helm-charts.newrelic.com
+  version: 0.15.0
```
You can update to 0.17.0

kang-makes commented on charts/newrelic-k8s-metrics-adapter/tests/deployment_test.yaml:
```diff
@@ -0,0 +1,106 @@
+      - hasDocuments:
+          count: 1
+        template: templates/deployment.yaml
+  - it: renders template with values for all helpers
+    set:
+      personalAPIKey: 21321
+      config:
+        accountID: 111
+        region: A-REGION
+      priorityClassName: high-priority
+      hostNetwork: true
+      dnsConfig:
+        keyOnValues: bar
+      tolerations:
+        - key: "key1"
+          operator: "Exists"
+          effect: "NoSchedule"
+      nodeSelector:
+        disktype: ssd
+      affinity:
+        nodeAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+              - matchExpressions:
+                  - key: kubernetes.io/e2e-az-name
+                    operator: In
+                    values:
+                      - new
+    asserts:
+      - hasDocuments:
+          count: 1
+        template: templates/deployment.yaml
```
NIT: I would remove this and add it to a file in `charts/newrelic-k8s-metrics-adapter/ci/whatever.yaml` so it would not only test that it would template but creates a valid manifest that is installable.

Sorry for do not think about this possibility earlier.

kang-makes commented on charts/newrelic-k8s-metrics-adapter/tests/job_patch_job_createsecret_test.yaml:
```diff
@@ -0,0 +1,56 @@
+      - matchRegex:
+          path: spec.template.spec.containers[0].image
+          pattern: ^.*k8s.gcr.io/ingress-nginx/kube-webhook-certgen:latest
+  - it: renders when extraEnv, extraEnvFrom, extraVolumes and extraVolumeMounts are set
+    set:
+      personalAPIKey: 21321
+      tolerations:
+        - key: "key1"
+          operator: "Exists"
+          effect: "NoSchedule"
+      apiServicePatchJob:
+        extraEnv:
+          - name: ENV_VAR1
+            value: "var1"
+          - name: ENV_VAR2
+            value: "var2"
+        extraEnvFrom:
+          - configMapRef:
+              name: special-config
+        extraVolumes:
+          - name: tmpfs-data
+            emptyDir: {}
+        extraVolumeMounts:
+          - mountPath: /var/db/newrelic-infra/data
+            name: tmpfs-data
+    asserts:
+      - hasDocuments:
+          count: 1
```
Same as above.

alvarocabanas replied:
done

kang-makes commented on charts/newrelic-k8s-metrics-adapter/tests/job_patch_job_patchapiservice_test.yaml:
```diff
@@ -0,0 +1,59 @@
+      - matchRegex:
+          path: spec.template.spec.containers[0].image
+          pattern: ^.*k8s.gcr.io/ingress-nginx/kube-webhook-certgen:latest
+  - it: renders when extraEnv, extraEnvFrom, extraVolumes and extraVolumeMounts are set
+    set:
+      personalAPIKey: 21321
+      tolerations:
+        - key: "key1"
+          operator: "Exists"
+          effect: "NoSchedule"
+      apiServicePatchJob:
+        extraEnv:
+          - name: ENV_VAR1
+            value: "var1"
+          - name: ENV_VAR2
+            value: "var2"
+        extraEnvFrom:
+          - configMapRef:
+              name: special-config
+        extraVolumes:
+          - name: tmpfs-data
+            emptyDir: {}
+        extraVolumeMounts:
+          - mountPath: /var/db/newrelic-infra/data
+            name: tmpfs-data
+    asserts:
+      - hasDocuments:
+          count: 1
```
Same as above.

kang-makes reviewed (approved):
I only left NITs. LGTM.

roobre commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/clusterrole.yaml:
```diff
@@ -2,13 +2,13 @@
   labels:
-    app: {{ template "newrelic-k8s-metrics-adapter.name" $ }}
-{{ include "newrelic-k8s-metrics-adapter.labels" $ | indent 4 }}
+    app: {{ template "common.naming.name" . }}
```
Should we keep the `app` label? `common.labels` already includes `app.kubernetes.io/name: {{ include "common.naming.name" . }}`

roobre commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/clusterrolebinding.yaml:
```diff
@@ -2,19 +2,19 @@
   labels:
-    app: {{ template "newrelic-k8s-metrics-adapter.name" $ }}
-{{ include "newrelic-k8s-metrics-adapter.labels" $ | indent 4 }}
+    app: {{ template "common.naming.name" . }}
```
Ditto

roobre commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/job-createSecret.yaml:
```diff
@@ -3,48 +3,60 @@ apiVersion: batch/v1
   labels:
-    app: {{ template "newrelic-k8s-metrics-adapter.name" $ }}
-{{ include "newrelic-k8s-metrics-adapter.labels" $ | indent 4 }}
+    app: {{ template "common.naming.name" . }}
```
Ditto

roobre replied:
(Just to avoid spam I won't ditto again but we have this label in many places)

roobre commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/job-createSecret.yaml:
```diff
@@ -3,48 +3,60 @@ apiVersion: batch/v1
             - --cert-name=tls.crt
             - --key-name=tls.key
-          {{- if .Values.apiServicePatchJob.image.volumeMounts }}
+          {{- with .Values.apiServicePatchJob.extraEnv }}
+          env:
+          {{- toYaml . | nindent 10 }}
+          {{- end }}
+          {{- with .Values.apiServicePatchJob.extraEnvFrom }}
+          envFrom:
+          {{- toYaml . | nindent 10 }}
+          {{- end }}
+          {{- with .Values.apiServicePatchJob.extraVolumeMounts }}
```
Do we need this snippet? As this is going to be a short-lived job I'm mildly reluctant to add features to it.

alvarocabanas replied:
I have removed the env and ExtraEnv and left only the Volumes and VolumeMounts since they were already present in the previous chart.

roobre commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/job-patchAPIService.yaml:
```diff
@@ -3,50 +3,58 @@ apiVersion: batch/v1
             - --apiservice-name=v1beta1.external.metrics.k8s.io
-          {{- if .Values.apiServicePatchJob.image.volumeMounts }}
+          {{- with .Values.apiServicePatchJob.extraEnv }}
+          env:
+          {{- toYaml . | nindent 10 }}
+          {{- end }}
+          {{- with .Values.apiServicePatchJob.extraEnvFrom }}
+          envFrom:
+          {{- toYaml . | nindent 10 }}
+          {{- end }}
+          {{- with .Values.apiServicePatchJob.extraVolumeMounts }}
```
As above

alvarocabanas replied:
same as above

roobre commented on charts/newrelic-k8s-metrics-adapter/templates/_helpers.tpl:
```diff
@@ -98,13 +13,9 @@ Select a value for the region
 {{- define "newrelic-k8s-metrics-adapter.region" -}}
```
From @paologallinaharbur's comments below I see that US is the default, but this might be non-obvious. Could we have a quick comment saying that?

roobre commented on charts/newrelic-k8s-metrics-adapter/values.yaml:
```diff
@@ -62,17 +63,18 @@ resources:
 
 serviceAccount:
   # Specifies whether a ServiceAccount should be created for the job and the deployment.
-  create: true
+  create:
   # The name of the ServiceAccount to use.
   # If not set and create is true, a name is generated using the fullname template.
   name:
```
Previously the defaults were explicit, but now they are not. Could we add a comment explaining them? Otherwise users might get confused.

alvarocabanas replied:
I've added the comments

roobre reviewed (commented):
LGTM, left some comments here and there. Awesome work so far!

roobre commented on charts/newrelic-k8s-metrics-adapter/templates/_helpers.tpl:
```diff
@@ -1,110 +1,22 @@
-{{- else if eq (include "newrelic-k8s-metrics-adapter.licenseKey" . | substr 0 2) "eu" -}}
+{{- else if (include "common.nrStaging" .)  -}}
+Staging
+{{- else if eq (include "common.license._licenseKey" . | substr 0 2) "eu" -}}
```
We could use `contains` or `hasPrefix` to check this http://masterminds.github.io/sprig/strings.html

alvarocabanas replied:
I used hasPrefix, thanks for the link, I'll added to my toolset

roobre commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/job-patchAPIService.yaml:
```diff
@@ -3,50 +3,48 @@ apiVersion: batch/v1
+            - --secret-name={{ template "common.naming.fullname" . }}-apiservice
             - --apiservice-name=v1beta1.external.metrics.k8s.io
-          {{- if .Values.apiServicePatchJob.image.volumeMounts }}
+          {{- with .Values.apiServicePatchJob.volumeMounts }}
```
No need to fix this, but in the future template directives should have the same indentation as if they where values

```suggestion
            {{- with .Values.apiServicePatchJob.volumeMounts }}
```

roobre commented on charts/newrelic-k8s-metrics-adapter/templates/apiservice/job-patch/serviceaccount.yaml:
```diff
@@ -3,11 +3,11 @@ apiVersion: v1
   annotations:
     "helm.sh/hook": pre-install,pre-upgrade,post-install,post-upgrade
     "helm.sh/hook-delete-policy": before-hook-creation,hook-succeeded
+    "helm.sh/hook-weight": "-1"
```
I'd add a comment here so we remember why we needed to add this. Possibly referencing https://github.com/helm/helm/pull/10787.

alvarocabanas replied:
done

roobre reviewed (approved):
Left some comments, please feel free to ignore the one about indentation.
